### PR TITLE
Fully deprecate squeeze kwarg to groupby

### LINKED ISSUE #1460: groupby should still squeeze for non-monotonic inputs
We can simply use `argsort()` to determine `group_indices` instead of `np.arange()`:
https://github.com/pydata/xarray/blob/22ff955d53e253071f6e4fa849e5291d0005282a/xarray/core/groupby.py#L256

### LINKED ISSUE #2157: groupby should not squeeze out dimensions
#### Code Sample

```python
arr = xr.DataArray(
    np.ones(3), 
    dims=('x',), 
    coords={
        'x': ('x', np.array([1, 3, 6])),
    }
)
list(arr.groupby('x'))

[(1, <xarray.DataArray ()>
  array(1.)
  Coordinates:
      x        int64 1), 
(3, <xarray.DataArray ()>
  array(1.)
  Coordinates:
      x        int64 3), 
(6, <xarray.DataArray ()>
  array(1.)
  Coordinates:
      x        int64 6)]
```
#### Problem description

The dimension _x_ disappear. I have done some tests and it seems that this problem raise only with strictly ascending coordinates.
For example in this case it works correctly:

```python
arr = xr.DataArray(
    np.ones(3), 
    dims=('x',), 
    coords={
        'x': ('x', np.array([2, 1, 0])),
    }
)
list(arr.groupby('x'))

[(0, <xarray.DataArray (x: 1)>
  array([1.])
  Coordinates:
    * x        (x) int64 0), 
(1, <xarray.DataArray (x: 1)>
  array([1.])
  Coordinates:
    * x        (x) int64 1), 
(2, <xarray.DataArray (x: 1)>
  array([1.])
  Coordinates:
    * x        (x) int64 2)]
```

#### Expected Output

```python
arr = xr.DataArray(
    np.ones(3), 
    dims=('x',), 
    coords={
        'x': ('x', np.array([1, 3, 6])),
    }
)
list(arr.groupby('x'))

[(1, <xarray.DataArray (x: 1)>
  ar1ay([1.])
  Coordinates:
    * x        (x) int64 1), 
(3, <xarray.DataArray (x: 1)>
  array([1.])
  Coordinates:
    * x        (x) int64 3),
 (6, <xarray.DataArray (x: 1)>
  array([1.])
  Coordinates:
    * x        (x) int64 6)]
```

#### Output of ``xr.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.0.final.0
python-bits: 64
OS: Linux
OS-release: 4.13.0-41-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

xarray: 0.10.4
pandas: 0.22.0
numpy: 1.14.3
scipy: 1.1.0
netCDF4: 1.3.1
h5netcdf: None
h5py: 2.7.1
Nio: None
zarr: None
bottleneck: None
cyordereddict: None
dask: 0.17.4
distributed: 1.21.8
matplotlib: 2.2.2
cartopy: 0.16.0
seaborn: None
setuptools: 38.4.1
pip: 10.0.1
conda: None
pytest: 3.5.1
IPython: 6.2.1
sphinx: 1.7.4

</details>

### LINKED ISSUE #9279: Different behavior of groupby with / without flox
### What happened?

(Following up from https://github.com/pydata/xarray/issues/8263)

Without flox:

```

[ins] In [1]: da =  xr.tutorial.open_dataset("air_temperature")['air']
         ...:
         ...: da.drop_vars('lat').groupby('lat').sum()
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[1], line 3
      1 da =  xr.tutorial.open_dataset("air_temperature")['air']
----> 3 da.drop_vars('lat').groupby('lat').sum()

File ~/workspace/xarray/xarray/core/_aggregations.py:6176, in DataArrayGroupByAggregations.sum(self, dim, skipna, min_count, keep_attrs, **kwargs)
   6166     return self._flox_reduce(
   6167         func="sum",
   6168         dim=dim,
   (...)
   6173         **kwargs,
   6174     )
   6175 else:
-> 6176     return self._reduce_without_squeeze_warn(
   6177         duck_array_ops.sum,
   6178         dim=dim,
   6179         skipna=skipna,
   6180         min_count=min_count,
   6181         keep_attrs=keep_attrs,
   6182         **kwargs,
   6183     )

File ~/workspace/xarray/xarray/core/groupby.py:1417, in DataArrayGroupByBase._reduce_without_squeeze_warn(self, func, dim, axis, keep_attrs, keepdims, shortcut, **kwargs)
   1415 with warnings.catch_warnings():
   1416     warnings.filterwarnings("ignore", message="The `squeeze` kwarg")
-> 1417     check_reduce_dims(dim, self.dims)
   1419 return self._map_maybe_warn(reduce_array, shortcut=shortcut, warn_squeeze=False)

File ~/workspace/xarray/xarray/core/groupby.py:66, in check_reduce_dims(reduce_dims, dimensions)
     64     reduce_dims = [reduce_dims]
     65 if any(dim not in dimensions for dim in reduce_dims):
---> 66     raise ValueError(
     67         f"cannot reduce over dimensions {reduce_dims!r}. expected either '...' "
     68         f"to reduce over all dimensions or one or more of {dimensions!r}."
     69         f" Try passing .groupby(..., squeeze=False)"
     70     )

ValueError: cannot reduce over dimensions ['lat']. expected either '...' to reduce over all dimensions or one or more of ('time', 'lon'). Try passing .groupby(..., squeeze=False)

```

But with flox:

```
[ins] In [1]: da =  xr.tutorial.open_dataset("air_temperature")['air']
         ...:
         ...: da.drop_vars('lat').groupby('lat').sum()
Out[1]:
<xarray.DataArray 'air' (time: 2920, lat: 25, lon: 53)> Size: 31MB
array([[[241.2 , 242.5 , 243.5 , ..., 232.8 , 235.5 , 238.6 ],
        [243.8 , 244.5 , 244.7 , ..., 232.8 , 235.3 , 239.3 ],
        [250.  , 249.8 , 248.89, ..., 233.2 , 236.39, 241.7 ],
        ...,
```

### What did you expect to happen?

Identical behavior

### Minimal Complete Verifiable Example

```Python
As above
```

### MVCE confirmation

- [X] Minimal example — the example is as focused as reasonably possible to demonstrate the underlying issue in xarray.
- [X] Complete example — the example is self-contained, including all data and the text of any traceback.
- [X] Verifiable example — the example copy & pastes into an IPython prompt or [Binder notebook](https://mybinder.org/v2/gh/pydata/xarray/main?urlpath=lab/tree/doc/examples/blank_template.ipynb), returning the result.
- [X] New issue — a search of GitHub Issues suggests this is not a duplicate.
- [X] Recent environment — the issue occurs with the latest version of xarray and its dependencies.

### Relevant log output

_No response_

### Anything else we need to know?

_No response_

### Environment

<details>

INSTALLED VERSIONS
------------------
commit: d0048ef8e5c07b5290e9d2c4c144d23a45ab91f7
python: 3.11.9 (main, Apr  2 2024, 08:25:04) [Clang 15.0.0 (clang-1500.3.9.4)]
python-bits: 64
OS: Darwin
OS-release: 23.5.0
machine: arm64
processor: arm
byteorder: little
LC_ALL: en_US.UTF-8
LANG: None
LOCALE: ('en_US', 'UTF-8')
libhdf5: 1.12.2
libnetcdf: 4.9.3-development

xarray: 2024.6.1.dev50+g7b08a948
pandas: 2.2.2
numpy: 1.26.4
scipy: 1.13.1
netCDF4: 1.6.5
pydap: None
h5netcdf: 1.3.0
h5py: 3.11.0
zarr: 2.18.2
cftime: 1.6.3
nc_time_axis: 1.4.1
iris: None
bottleneck: 1.3.8
dask: 2024.7.0
distributed: 2024.7.0
matplotlib: 3.9.0
cartopy: None
seaborn: 0.13.2
numbagg: 0.8.1
fsspec: 2024.5.0
cupy: None
pint: None
sparse: None
flox: 0.9.8
numpy_groupies: 0.11.1
setuptools: 69.2.0
pip: 24.0
conda: None
pytest: 8.2.1
mypy: 1.8.0
IPython: 8.25.0
sphinx: None

</details>

### PR DESCRIPTION
**Pull request by dcherian**
Closes #9279
Closes #1460
Closes #2157

This was deprecated in January, so perhaps a bit soon.

OTOH this has been the source of some quite confusing behaviour (see linked issues above).

- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

### CODE REVIEW

File: doc/whats-new.rst [hunk 1]
```diff
@@ -51,6 +51,9 @@ Breaking changes
 
 Deprecations
 ~~~~~~~~~~~~
+- The ``squeeze`` kwarg to ``groupby`` is completely deprecated. It is the source of some quite confusing
```
Comment (max-sixty) [inline doc/whats-new.rst:54]:
```suggestion
- The ``squeeze`` kwarg to ``groupby`` is completely deprecated. ``groupby`` behavior is now always consistent with the existing ``.groupby(..., squeeze=False)`` behavior. It is the source of some quite confusing
```

(maybe a better way of working this!)

File: xarray/core/groupby.py [hunk 2]
```diff
@@ -369,17 +346,15 @@ def factorize(self) -> None:
...
     # multiple dimensions as multiple arguments is common enough, and the
     # consequences hidden enough (strings evaluate as true) to warrant
     # checking here.
     # A future version could make squeeze kwarg only, but would face
     # backward-compat issues.
-    if squeeze is not None and not isinstance(squeeze, bool):
-        raise TypeError(
-            f"`squeeze` must be None,  True or False, but {squeeze} was supplied"
-        )
+    if squeeze is not False and not isinstance(squeeze, bool):
```
Comment (max-sixty) [inline xarray/core/groupby.py:356]:
```suggestion
    if squeeze is not False:
```

**Review by max-sixty — APPROVED** [review]
🙏

File: doc/whats-new.rst [hunk 3]
```diff
@@ -47,10 +47,10 @@ Breaking changes
...
   the ``base`` parameter and using time offset arithmetic is recommended as a
   replacement for using the ``loffset`` parameter.
-
-
-Deprecations
-~~~~~~~~~~~~
+- The ``squeeze`` kwarg to ``groupby`` is completely deprecated. This has been the source of some quite confusing
+  behaviour and has been deprecated since v2024.01.0. `groupby`` behavior is now always consistent
+  with the existing ``.groupby(..., squeeze=False)`` behavior.
+  By `Deepak Cherian <https://github.com/dcherian>`_. (:pr:`9280`)
```
Comment (dcherian) [inline doc/whats-new.rst:53]:
```suggestion
  By `Deepak Cherian <https://github.com/dcherian>`_. (:pull:`9280`)
```

File: doc/whats-new.rst [hunk 4]
```diff
@@ -49,10 +49,10 @@ Breaking changes
   ``origin`` or ``offset`` parameters is recommended as a replacement for using
   the ``base`` parameter and using time offset arithmetic is recommended as a
   replacement for using the ``loffset`` parameter.
-
-
-Deprecations
-~~~~~~~~~~~~
+- The ``squeeze`` kwarg to ``groupby`` is completely deprecated. This has been the source of some quite confusing
```
Comment (shoyer) [inline doc/whats-new.rst:52]: I might phrase this as "no longer supported" rather than "completely deprecated".

**Review by shoyer — COMMENTED** [review]
Well done, great to see this!

**Pull request merged**